include/swift/AST/AutoDiff.h

```diff
@@ -148,10 +168,59 @@ void getSubsetParameterTypes(IndexSubset *indices, AnyFunctionType *type,
 
 namespace llvm {
 
+using swift::AutoDiffConfig;
 using swift::AutoDiffDerivativeFunctionKind;
+using swift::GenericSignature;
+using swift::IndexSubset;
 
 template <typename T> struct DenseMapInfo;
 
+template <> struct DenseMapInfo<AutoDiffConfig> {
+  static AutoDiffConfig getEmptyKey() {
+    auto *ptr = llvm::DenseMapInfo<void *>::getEmptyKey();
+    // The `derivativeGenericSignature` component must be `nullptr` so that
```

Comment:
Note: `GenericSignatureImpl::getCanonicalSignature()` crashes for `GenericSignature` with underlying `nullptr`, which seems like an anti-pattern. I'm not sure whether this is intentional, so I [posted a forum question](https://forums.swift.org/t/make-genericsignature-getcanonicalsignature-not-crash-for-nullptr-genericsignature/32409?u=dan-zheng).

Reply:
@jckarter suggested defining `GenericSignature::getCanonicalSignature()` with the expected behavior as a fix. I'll investigate that in a follow-up.

Reply:
`GenericSignature::getCanonicalSignature` done in https://github.com/apple/swift/pull/29105.

However, I'd like to defer rewriting uses of `GenericSignatureImpl::getCanonicalSignature` in `AutoDiffConfig` until changes have been downstreamed to and tested on `tensorflow` branch.